### ISSUE TITLE
Need a way to detect cascade restrict exceptions

### ISSUE THREAD

**Issue by sjb-sjb** [dotnet/efcore#8200]
@rowanmiller, consistent with your comment in #4628, I have set cascade restrict whenever the foreignKey.IsRequired because in my application I normally do not want to delete any parent records that are required by dependent records. 

What I would like to do is catch the cascade restrict exceptions and inform the end user that the (parent) record cannot be deleted because it is in use. Is there any clean way to identify the cascade restrict exceptions -- i.e. other than by parsing the error message that you cited in https://www.safaribooksonline.com/library/view/programming-entity-framework/9781449317867/ch04s04.html ?  Ideally the code would be maintainable independent of the local language and preferably also independent of the database provider. As it happens I am localizing in a couple of languages, and using SQLite but potentially also SqlServer.  

The only alternative I see to identifying the cascade restrict exceptions would be to do a query for the dependents (or for the count of them). I would have to check to see if the entity state is deleted, then get the child navigation properties, check the delete behavior (maybe of the inverse navigation?) to see if it is restrict, extract the query for the child navigation property and modify it to do a count, and execute the modified query. Presumably the count and the delete operation would also have to be in a transaction in order to prevent a change to the number of child entities between the count and the delete. It would be very inconvenient (especially the transaction) and also not particularly good from a performance point of view.

Thoughts on this?

**Comment by ajcvickers** [comment]
@sjb-sjb I think you need to consider what is happening where. In general, EF can never tell you if there is going to be a constraint violation when saving changes. This is because EF never knows exactly what is in the database. That being said, if you use a transaction and have EF load all dependent entities, recursively, then you would be able to make an assessment as to whether or not a dependent is going to cause a constraint violation. But I think this would be taking the wrong approach. Instead, it seems like it is the job of your business logic to ensure that only allowed operations happen on the data, with appropriate UI and feedback to the users to enforce this. The database constraint is then just a way of catching errors in your business logic. Because without this, how do you even communicate to a user that they did something wrong in a way that makes sense to them? Just saying that they can't delete this row because some other row is referencing it will likely be meaningless. And if it isn't meaningless, isn't that what the message from the database will tell them anyway?

**Comment by sjb-sjb** [comment]
@ajcvickers: Philosophically I agree with the point that it is business logic rather than database triggers that should drive your ui. However this does not solve the problem because the business logic is well represented by the object design.  A required relationship and cascade  restrict represents the business logic that the parent entity cannot be deleted if the child entity is present.

Let me give an example. Say I am writing a product management system. I have Products each of which has a Description, an amount of Inventory and/or a number of Backorders. The Product is the parent record and the others are children. Business logic is that if you delete the Product then the Description can also go, so we set cascade delete there, but you cannot delete a Product while there is Inventory or Backorder, so we set cascade Restrict there. 

Now it's true that we need a localized message associating the possible reasons ("inventory is on hand or backorders are open") with the Product entity. Other than that, though, the problem is what I described: either test the count of child records and delete the parent in a transaction, or else catch the cascade restrict exception. I am hoping there is an easy way to do it. 

As for catching the exception, if I do have to check the text then is this dependent on the language of the system or the user's localization settings? Can I be sure of the dB version being used, for example if I am using  SQLite under UWP is this something that may vary depending on the user's machine configuration or is it determined by my UWP app?

Either approach seems painful. 

- sjb

P.S. I liked your comment about having to check recursively. This doesn't happen very often but I would count this as a point in favor of catching the exception.
